src/librustc_mir/transform/move_up_propagation.rs

```diff
@@ -0,0 +1,496 @@
+            let new_stmts: Vec<_> = bbs[bb].statements
+                                            .iter()
+                                            .enumerate()
+                                            .map(|(stmt_idx, orig_stmt)| {
+                if let Some(repl) = repls.get(&stmt_idx) {
+                    let repl_stmt = Statement { kind: repl.clone(),
+                                                source_info: orig_stmt.source_info,
+                    };
+                    debug!("replacing {:?} with {:?}", orig_stmt, repl_stmt);
+                    repl_stmt
+                } else {
+                    orig_stmt.clone()
+                }
+            }).collect();
+            bbs[bb] = BasicBlockData {
+                statements: new_stmts,
+                terminator: bbs[bb].terminator.clone(),
+                is_cleanup: bbs[bb].is_cleanup,
+            };
+        }
+
+        for (bb, dead_idxs) in dead {
+            let mut idx_cnt = 0;
+            bbs[bb].statements.retain(|_| {
+                let dead = dead_idxs.contains(idx_cnt);
+                idx_cnt += 1;
+                !dead
+            });
+        }
+    }
+}
+
+fn paths_satisfy_our_condition<'a>(dest: &Lvalue<'a>,
+                                   end_statement: &Statement<'a>,
+                                   end_block: BasicBlock,
+                                   start_block: BasicBlock,
+                                   start_index: usize,
+                                   start_block_data: &BasicBlockData<'a>)
+                                   -> bool {
+    let mut upf = UseOnPathFinder {
+        dest: dest,
+        end_statement: end_statement,
+        end_block: end_block,
+        found_intermediate_use_of_dest: false,
+        found_call: false,
+        found_end_statement: false,
+    };
+    upf.visit_first_block(start_block, start_block_data, start_index);
+    upf.found_end_statement && !upf.found_intermediate_use_of_dest && !upf.found_call
+}
+
+struct UseOnPathFinder<'a> {
+    dest: &'a Lvalue<'a>,
+    end_statement: &'a Statement<'a>,
+    end_block: BasicBlock,
+    found_intermediate_use_of_dest: bool,
+    found_call: bool,
+    found_end_statement: bool,
+}
+
+impl<'a> UseOnPathFinder<'a> {
+    fn visit_first_block(&mut self,
+                         block: BasicBlock,
+                         data: &BasicBlockData<'a>,
+                         start_index: usize) {
+
+        for (i, statement) in data.statements.iter().enumerate() {
+            if i > start_index {
+                self.visit_statement(block, statement);
+            }
+        }
+
+        if let Some(ref terminator) = data.terminator {
+            self.visit_terminator(block, terminator);
+        }
+    }
+}
+
+impl<'a> Visitor<'a> for UseOnPathFinder<'a> {
+    fn visit_statement(&mut self, block: BasicBlock, statement: &Statement<'a>) {
+        if self.end_block == block && self.end_statement == statement {
```

Comment:
I think I agree with @arielb1 that it'd be better to use indices than references (and we could probably remove the `PartialEq` impls).